### PR TITLE
Update TMVA BDT parallelisation PR. This is un update of PR #723 

### PR DESCRIPTION
**Pull request by lmoneta**
Update and replace PR #723 fixing the conflict with master

### CODE REVIEW

**Comment by lmoneta** [comment]
@phsft-bot build

File: tmva/tmva/src/DecisionTree.cxx [hunk 1]
```diff
@@ -359,9 +716,9 @@ UInt_t TMVA::DecisionTree::BuildTree( const std::vector<const TMVA::Event*> & ev
 
    if (s+b < 0) {
       Log() << kWARNING << " One of the Decision Tree nodes has negative total number of signal or background events. "
-            << "(Nsig="<<s<<" Nbkg="<<b<<" Probably you use a Monte Carlo with negative weights. That should in principle "
+            << "(Nsig="<<s<<" Nbkg="<<b<<" Probaby you use a Monte Carlo with negative weights. That should in principle "
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: This introduces a spelling mistake, which should probably be avoided ;)

File: tmva/tmva/src/DecisionTree.cxx [hunk 2]
```diff
@@ -359,9 +716,9 @@ UInt_t TMVA::DecisionTree::BuildTree( const std::vector<const TMVA::Event*> & ev
...
             << "be fine as long as on average you end up with something positive. For this you have to make sure that the "
-            << "minimal number of (unweighted) events demanded for a tree node (currently you use: MinNodeSize="<<fMinNodeSize
+            << "minimul number of (unweighted) events demanded for a tree node (currently you use: MinNodeSize="<<fMinNodeSize
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Same as above

File: tmva/tmva/src/DecisionTree.cxx [hunk 3]
```diff
@@ -400,10 +760,10 @@ UInt_t TMVA::DecisionTree::BuildTree( const std::vector<const TMVA::Event*> & ev
    // of events in the parent node is not at least two times as big, I don't even need to try
    // splitting
 
-   // ask here for actual "events" independent of their weight.. OR the weighted events
-   // to exceed the min requested number of events per daughter node
+   // ask here for actuall "events" independent of their weight.. OR the weighted events
+   // to execeed the min requested number of events per dauther node
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Spelling mistake introduced.

File: tmva/tmva/src/DecisionTree.cxx [hunk 4]
```diff
@@ -400,10 +760,10 @@ UInt_t TMVA::DecisionTree::BuildTree( const std::vector<const TMVA::Event*> & ev
...
    // (NOTE: make sure that at the eventSample at the ROOT node has sum_of_weights == sample.size() !
-   //   if ((eventSample.size() >= 2*fMinSize ||s+b >= 2*fMinSize) && node->GetDepth() < fMaxDepth
+   //   if ((eventSample.size() >= 2*fMinSize ||s+b >= 2*fMinSize) && node->GetDepth() < fMaxDepth
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Remove extra space at end of line.

File: tmva/tmva/src/DecisionTree.cxx [hunk 5]
```diff
@@ -918,6 +1279,100 @@ void TMVA::DecisionTree::GetRandomisedVariables(Bool_t *useVariable, UInt_t *map
    if (nSelectedVars != useNvars) { std::cout << "Bug in TrainNode - GetRandisedVariables()... sorry" << std::endl; std::exit(1);}
 }
 
+// Multithreaded version of DecisionTree::TrainNodeFast
+#ifdef R__USE_IMT
+//====================================================================================
+// Had to define this struct to enable parallelization in TrainNodeFast.
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: This comment should be rewritten to be more general (describe intent of the struct) and using doxygen documentation style.

File: tmva/tmva/src/DecisionTree.cxx [hunk 6]
```diff
@@ -918,6 +1279,100 @@ void TMVA::DecisionTree::GetRandomisedVariables(Bool_t *useVariable, UInt_t *map
...
+       ret.nTotS = nTotS + other.nTotS;
+       ret.nTotS_unWeighted = nTotS_unWeighted + other.nTotS_unWeighted;
+       ret.nTotB = nTotB + other.nTotB;
+       ret.nTotB_unWeighted = nTotB_unWeighted + other.nTotB_unWeighted;
+
+       return ret;
+   };
+ 
+};
+//===========================================================================
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Remove

File: tmva/tmva/src/DecisionTree.cxx [hunk 7]
```diff
@@ -930,18 +1385,19 @@ void TMVA::DecisionTree::GetRandomisedVariables(Bool_t *useVariable, UInt_t *map
 Double_t TMVA::DecisionTree::TrainNodeFast( const EventConstList & eventSample,
                                             TMVA::DecisionTreeNode *node )
 {
-   Double_t  separationGainTotal = -1, sepTmp;
+   // #### OK let's comment this one to see how to parallelize it
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: ???, should this comment be removed?

File: tmva/tmva/src/DecisionTree.cxx [hunk 8]
```diff
@@ -1125,34 +1582,571 @@ Double_t TMVA::DecisionTree::TrainNodeFast( const EventConstList & eventSample,
...
+         // now scan trough the cuts for each varable and find which one gives
+         // the best separationGain at the current stage.
+         if ( useVariable[ivar] ) {
+            Double_t eventData;
+            if (ivar < fNvars) eventData = eventSample[iev]->GetValue(ivar);
             else { // the fisher variable
                eventData = fisherCoeff[fNvars];
                for (UInt_t jvar=0; jvar<fNvars; jvar++)
-                  eventData += fisherCoeff[jvar]*(eventSample[iev])->GetValueFast(jvar);
+                  eventData += fisherCoeff[jvar]*(eventSample[iev])->GetValue(jvar);
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Should this really be changed to GetValue?

File: tmva/tmva/src/DecisionTree.cxx [hunk 9]
```diff
@@ -1199,19 +2194,20 @@ Double_t TMVA::DecisionTree::TrainNodeFast( const EventConstList & eventSample,
...
-   // now select the optimal cuts for each variable and find which one gives
+   // #### Loops over vars and bins, but not events, not worth parallelizing unless nbins is on the order of ndata/10 ish ...
+   // now select the optimal cuts for each varable and find which one gives
    // the best separationGain at the current stage
    for (UInt_t ivar=0; ivar < cNvars; ivar++) {
       if (useVariable[ivar]) {
          for (UInt_t iBin=0; iBin<nBins[ivar]-1; iBin++) { // the last bin contains "all events" -->skip
             // the separationGain is defined as the various indices (Gini, CorssEntropy, e.t.c)
-            // calculated by the "SamplePurities" from the branches that would go to the
+            // calculated by the "SamplePurities" fom the branches that would go to the
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Spelling mistake

File: tmva/tmva/src/DecisionTree.cxx [hunk 10]
```diff
@@ -1199,19 +2194,20 @@ Double_t TMVA::DecisionTree::TrainNodeFast( const EventConstList & eventSample,
...
             // left or the right from this node if "these" cuts were used in the Node:
             // hereby: nSelS and nSelB would go to the right branch
             //        (nTotS - nSelS) + (nTotB - nSelB)  would go to the left branch;
 
-            // only allow splits where both daughter nodes match the specified minimum number
-            // for this use the "unweighted" events, as you are interested in statistically
+            // only allow splits where both daughter nodes match the specified miniumum number
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Spelling mistake

File: tmva/tmva/src/MethodBDT.cxx [hunk 11]
```diff
@@ -1316,18 +1330,21 @@ void TMVA::MethodBDT::Train()
...
          if (fUseFisherCuts) {
             fForest.back()->SetUseFisherCuts();
             fForest.back()->SetMinLinCorrForFisher(fMinLinCorrForFisher);
             fForest.back()->SetUseExclusiveVars(fUseExclusiveVars);
          }
-
+         
          nNodesBeforePruning = fForest.back()->BuildTree(*fTrainSample);
-
+
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBDT.cxx]: Unecessary whitespace

File: tmva/tmva/src/MethodBDT.cxx [hunk 12]
```diff
@@ -1525,6 +1565,7 @@ Double_t TMVA::MethodBDT::GradBoostRegression(std::vector<const TMVA::Event*>& e
 {
    // get the vector of events for each terminal so that we can calculate the constant fit value in each
    // terminal node
+   // #### Not sure how many events are in each node in advance, so I can't parallelize this easily
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBDT.cxx]: This is maybe better put in external document?

File: tmva/tmva/src/MethodBDT.cxx [hunk 13]
```diff
@@ -1533,13 +1574,16 @@ Double_t TMVA::MethodBDT::GradBoostRegression(std::vector<const TMVA::Event*>& e
 
    // calculate the constant fit for each terminal node based upon the events in the node
    // node (iLeave->first), vector of event information (iLeave->second)
+   // #### could parallelize this and do the leaves at the same time, but this doesn't take very long compared
+   // #### to the other processes
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBDT.cxx]: This is maybe better put in external document?

File: tmva/tmva/src/DecisionTree.cxx [hunk 14]
```diff
@@ -348,8 +704,9 @@ UInt_t TMVA::DecisionTree::BuildTree( const std::vector<const TMVA::Event*> & ev
          target2+=weight*tgt*tgt;
       }
 
+      // save the min and max for each feature
       for (UInt_t ivar=0; ivar<fNvars; ivar++) {
-         const Double_t val = evt->GetValueFast(ivar);
+         const Double_t val = evt->GetValue(ivar);
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Should this be changed?

File: tmva/tmva/src/DecisionTree.cxx [hunk 15]
```diff
@@ -1125,34 +1582,571 @@ Double_t TMVA::DecisionTree::TrainNodeFast( const EventConstList & eventSample,
...
+            }
+
+            // #### Count the number in each bin
+            Int_t iBin=-1;
+            for (UInt_t ivar=0; ivar < cNvars; ivar++) {
+               // now scan trough the cuts for each varable and find which one gives
+               // the best separationGain at the current stage.
+               if ( useVariable[ivar] ) {
+                  Double_t eventData;
+                  if (ivar < fNvars) eventData = eventSample[iev]->GetValue(ivar);
```
Comment (ashlaban) [inline tmva/tmva/src/DecisionTree.cxx]: Should this be changed from GetValueFast?

**Comment by lmoneta** [comment]
@phsft-bot -Dimt=OFF

**Comment by lmoneta** [comment]
@phsft-bot build -Dimt=OFF

**Comment by lmoneta** [comment]
@phsft-bot build with flags -Dimt=Off

**Pull request merged**